proposals/v3/api.md

```diff
@@ -0,0 +1,1266 @@
+```
+GET http://HOST:5000/synse/v3/transaction[/{transaction_id}]
+```
+
+Check the state and status of a write transaction.
+
+If no *transaction_id* is given, a sorted list of all cached transaction IDs is returned. The
+length of time that a transaction is cached is configurable (see the Synse Server [configuration](server.md#configuration)).
+
+If the provided transaction ID does not exist, an error is returned.
+
+#### URI Parameters
+
+| Parameter | Required | Description |
+| :-------- | :------- | :---------- |
+| *transaction_id* | no | The ID of the write transaction to get the status of. Transaction IDs are provided from an [asynchronous write](#write-asynchronous) response. |
+
+
+#### Response
+
+##### OK (200)
+```json
+{
+  "id": "56a32eba-1aa6-4868-84ee-fe01af8b2e6b",
+  "timeout": "10s",
+  "device": "0fe8f06229aa9a01ef6032d1ddaf18a5",
+  "context": {
+    "action": "color",
+    "data": "f38ac2"
+  },
+  "status": "done",
+  "created": "2018-02-01T15:00:51Z",
+  "updated": "2018-02-01T15:00:51Z",
+  "message": ""
+}
+```
+
+**Fields**
+
+| Field | Description |
+| :---- | :---------- |
+| *id* | The ID of the transaction. This is generated and assigned by the plugin being written to. |
+| *timeout* | A string representing the timeout for the write transaction after which it will be cancelled. This is effectively the maximum wait time for the transaction to resolve. |
+| *device* | The GUID of the device being written to. |
+| *context* | The POSTed write data for the given write transaction. |
+| *status* | The current status of the transaction. (`pending`, `writing`, `done`, `error`) |
+| *created* | The time at which the transaction was created. This timestamp is generated by the plugin. |
+| *updated* | The last time the transaction state *or* status was updated. Once the transaction reaches a `done` status or `error` state, no further updates will occur. |
+| *message* | Any context information relating to a transaction's error state. If there is no error, this will be an empty string. |
+
+
+Below is an example of the 200 OK response from the `/transaction` endpoint when no transaction ID
+is provided:
+
+```json
+[
+  "56a32eba-1aa6-4868-84ee-fe01af8b2e6b",
+  "56a32eba-1aa6-4868-84ee-fe01af8b2e6c",
+  "56a32eba-1aa6-4868-84ee-fe01af8b2e6d"
+]
+```
+
+##### Error
+See: [Errors](#errors)
+
+* **500** - Catchall processing error
+* **404** - Transaction not found
+
+#### Status
+Transaction can have one of four statuses:
+
+| Status | Description |
+| :----- | :---------- |
+| *pending* | The write was received and queued but has not yet been processed. |
+| *writing* | The write has been taken off the queue and is being processed. |
+| *done* | The write has been processed and completed successfully. (terminal state) |
+| *error* | The write failed for any reason. (terminal state) |
+
+
+---
+
+### Device
+```
+GET  http://HOST:5000/synse/v3/device/{device_id}
+POST http://HOST:5000/synse/v3/device/{device_id}
```

Comment:
An alternative to this could be to restructure the API a little bit so instead of having `/write/{device_id}`, we would
have `/{device_id}/write`. 

With this approach, we could have:
* `/device/{device_id}/read` - this would replace the 'read device' route `/read/{device_id}`
* `/device/{device_id}/write` - this would replace the 'write' route(s) `/write/{device_id}`
* `/device/{device_id}/info` - this would replace the info route `/info/{device_id}`

Those are essentially the only routes that target a single device, so others wouldn't necessarily need to be updated. There are other possible ways of structuring as well.. this is just a thought, as it allows us to tie `/info` off of a static device URI (for FQDN devices).

Reply:
I would prefer it as written, using RESTful endpoints instead of `/device/{id}/read`, etc.

Reply:
WRT `/read` `/write` VS. `/device` (GET, POST):

> would you recommend removing the /read/{device} and /write/{device} endpoints in favor of the /device/{device} endpoint?

I was definitely under the impression that they were both going to be supported. My opinion was against this suggestion above ^^^ where `device_id` gets stuffed in the middle of the URI. I support keeping a `/read` & `/write` endpoint, with support for a single device_id, or batch of tags. As well as a `/device/{device_id}` endpoint that responds differently to GET, and POST requests. It appears the `/device` endpoint is a result of [discussion](https://github.com/vapor-ware/synse-server/pull/231/files#r249986998) with @MatthewHink. All of my use cases are satisfied by `/read` & `/write`, so the question is whether or not `/device/{device_id}` satisfies the requirements for FQDN. Which it should, right? If the Device IDs are globally unique and deterministic. 
To answer your question above: No, I like both.

Reply:
👍 

I'm also happy leaving as-is, just wanted to propose alternatives in case others found those to be more useful. 

Yeah, the only real requirement I got for FQDN for individual devices was that they had to be uniquely addressable, so the `/device/{device_id}` endpoint fulfills that requirement since device IDs will be globally unique.

Thanks for the feedback @KylerBurke  and @marcoceppi  (: